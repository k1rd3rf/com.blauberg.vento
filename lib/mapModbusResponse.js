'use strict';

const { Parameter } = require('blaubergventojs');
const { getEnumKeyByEnumValue } = require('./mapEnum');

module.exports = (result) => {
  if (result != null) {
    const values = result.packet._dataEntries.reduce((acc, { parameter, value }) => {
      const key = getEnumKeyByEnumValue(Parameter, parameter) || parameter;
      return key ? { ...acc, [key]: value } : acc;
    }, {});

    let unittypelabel = 'Vento Expert';
    switch (values['UNIT_TYPE'][0]) {
      case 1:
        unittypelabel = 'Vento Expert A50-1 W V.2 | Vento Expert A85-1 W V.2 | Vento Expert A100-1 W V.2';
        break;
      case 4:
        unittypelabel = 'Vento Expert Duo A30-1 W V.2';
        break;
      case 5:
        unittypelabel = 'Vento Expert A30 W V.2';
        break;
      default:
        unittypelabel = 'Vento Expert';
        break;
    }

    return {
      onoff: values['ON_OFF'][0],
      speed: {
        mode: values['SPEED'][0],
        manualspeed: values['MANUAL_SPEED'][0],
      },
      boost: {
        mode: values['BOOT_MODE'][0],
        deactivationtimer: values['BOOST_MODE_DEACTIVATION_DELAY'][0],
      },
      operationmode: values['VENTILATION_MODE'][0],
      filter: {
        alarm: values['FILTER_ALARM'][0],
        timer: {
          min: values['FILTER_TIMER'][0],
          hour: values['FILTER_TIMER'][1],
          days: values['FILTER_TIMER'][2],
        },
      },
      humidity: {
        current: values['CURRENT_HUMIDITY'][0],
        sensoractivation: values['HUMIDITY_SENSOR_ACTIVATION'][0],
        threshold: values['HUMIDITY_THRESHOLD'][0],
        activated: 0,
      },
      unittype: unittypelabel,
      fan: {
        rpm: values['FAN1RPM'][0],
      },
      timers: {
        mode: values['TIMER_MODE'][0],
        countdown: {
          sec: values[11][0],
          min: values[11][1],
          hour: values[11][2],
        },
      },
      alarm: values['READ_ALARM'][0],
    };
  }
  throw new Error('device not responding, is your device password correct?');
};
