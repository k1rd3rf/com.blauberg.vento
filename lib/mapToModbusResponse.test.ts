import mapModbusResponse from './mapModbusResponse';

const result = {
  packet: {
    _deviceId: 'devId',
    _password: '1234',
    _functionType: 6,
    _dataEntries: [
      { parameter: 1, value: { 0: 1 } },
      { parameter: 2, value: { 0: 1 } },
      { parameter: 68, value: { 0: 179 } },
      { parameter: 6, value: { 0: 0 } },
      { parameter: 102, value: { 0: 30 } },
      { parameter: 183, value: { 0: 2 } },
      { parameter: 136, value: { 0: 0 } },
      {
        parameter: 100,
        value: {
          0: 34, 1: 20, 2: 31, 3: 0,
        },
      },
      { parameter: 37, value: { 0: 45 } },
      { parameter: 15, value: { 0: 1 } },
      { parameter: 25, value: { 0: 70 } },
      { parameter: 185, value: { 0: 4, 1: 0 } },
      { parameter: 74, value: { 0: 236, 1: 4 } },
      { parameter: 7, value: { 0: 0 } },
      { parameter: 131, value: { 0: 0 } },
      { parameter: 11, value: { 0: 0, 1: 0, 2: 0 } }],
  },
  ip: '127.0.0.1',
};

describe('mapToModbusResponse', () => {
  it('maps packet to Modbus response', () => {
    expect(mapModbusResponse(result)).toMatchSnapshot();
  });
});
