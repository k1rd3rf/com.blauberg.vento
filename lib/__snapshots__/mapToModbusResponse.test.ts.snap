// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`mapToModbusResponse maps packet to Modbus response 1`] = `
{
  "alarm": 0,
  "boost": {
    "deactivationtimer": 30,
    "mode": 0,
  },
  "fan": {
    "rpm": 236,
  },
  "filter": {
    "alarm": 0,
    "timer": {
      "days": 31,
      "hour": 20,
      "min": 34,
    },
  },
  "humidity": {
    "activated": 0,
    "current": 45,
    "sensoractivation": 1,
    "threshold": 70,
  },
  "onoff": 1,
  "operationmode": 2,
  "speed": {
    "manualspeed": 179,
    "mode": 1,
  },
  "timers": {
    "countdown": {
      "hour": 0,
      "min": 0,
      "sec": 0,
    },
    "mode": 0,
  },
  "unittype": "Vento Expert Duo A30-1 W V.2",
}
`;

exports[`mapToModbusResponse maps packet to Modbus response when device is off 1`] = `
{
  "alarm": 0,
  "boost": {
    "deactivationtimer": 30,
    "mode": 0,
  },
  "fan": {
    "rpm": 236,
  },
  "filter": {
    "alarm": 0,
    "timer": {
      "days": 31,
      "hour": 20,
      "min": 34,
    },
  },
  "humidity": {
    "activated": 0,
    "current": 45,
    "sensoractivation": 1,
    "threshold": 70,
  },
  "onoff": 0,
  "operationmode": 2,
  "speed": {
    "manualspeed": 179,
    "mode": 1,
  },
  "timers": {
    "countdown": {
      "hour": 0,
      "min": 0,
      "sec": 0,
    },
    "mode": 0,
  },
  "unittype": "Vento Expert Duo A30-1 W V.2",
}
`;
